### PR TITLE
Hook manager continuation

### PR DESCRIPTION
**Pull request by arielalmendral**
First draft of hook manager with one hook.

1. The QC workflow lives a parallel life together with a generic hook workflow, for backward compatibility. The QC hook is now called (in the code) a post_hook_workflow. This hook has by definition the run mode = POST_SIMULATION. Previous functionality from QC workflow is supported, please let me know if some extended test fail. 

2. The hook workflow can be invoked from the NEW config keyword with either:

HOOK_WORKFLOW /path/to/hook_workflow.py PRE_SIMULATION

or

HOOK_WORKFLOW /path/to/hook_workflow.py POST_SIMULATION

3. A PRE_SIMULATION hook will be triggered immediately after starting a "Simulation". The POST_SIMULATION hook is triggered after the simulation, just as with the QC workflow.

4. Smoother, Iterative Smoother, Experiment, ES_MDA and Simulation support a hook.

5. Application: A local config workflow is typically a PRE_SIMULATION hook.

### CODE REVIEW

**Comment by joakim-hove** [comment]
OK - I am looking into this; comments coming later today or tomorrow. Sorry it has taken so long.

File: devel/libenkf/src/hook_workflow.c [hunk 1]
```diff
@@ -0,0 +1,224 @@
+/*
+   Copyright (C) 2012  Statoil ASA, Norway.
```
Comment (joakim-hove) [inline devel/libenkf/src/hook_workflow.c]: 2015

**Comment by arielalmendral** [comment]
Great, thanks!

File: devel/libenkf/src/hook_workflow.c [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+#define QC_WORKFLOW_NAME    "QC WORKFLOW"
+#define RUNPATH_LIST_FILE   ".ert_runpath_list"
+
+
+struct hook_workflow_struct {
+  hook_run_mode_enum       run_mode;
+  char                   * path;
+  workflow_type          * workflow;
+  ert_workflow_list_type * workflow_list;
+  runpath_list_type      * runpath_list;
```
Comment (joakim-hove) [inline devel/libenkf/src/hook_workflow.c]: There should only be one `runpath_list` instance - which should be managed by the `hook_manager`.

File: devel/libenkf/src/hook_workflow.c [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+};
+
+
+hook_workflow_type * hook_workflow_alloc( ert_workflow_list_type * workflow_list , const char * path ) {
+  hook_workflow_type * hook_workflow = util_malloc( sizeof * hook_workflow );
+  hook_workflow->run_mode = PRE_SIMULATION;
```
Comment (joakim-hove) [inline devel/libenkf/src/hook_workflow.c]: The manager should handle `run_mode` -  it should not be necessary to have that all the way down here?

File: devel/libenkf/src/hook_workflow.c [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+#define RUN_MODE_POST_SIMULATION_NAME    "POST_SIMULATION"
+
+#define HOOK_WORKFLOW_NAME  "HOOK WORKFLOW"
+#define QC_WORKFLOW_NAME    "QC WORKFLOW"
+#define RUNPATH_LIST_FILE   ".ert_runpath_list"
+
+
+struct hook_workflow_struct {
+  hook_run_mode_enum       run_mode;
+  char                   * path;
```
Comment (joakim-hove) [inline devel/libenkf/src/hook_workflow.c]: What is this `path` variable?

File: devel/libenkf/include/ert/enkf/hook_workflow.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+extern "C" {
+#endif
+
+#include <ert/config/config_parser.h>
+
+#include <ert/enkf/ert_workflow_list.h>
+#include <ert/enkf/runpath_list.h>
+
+
+  typedef enum {PRE_SIMULATION  = 0,
```
Comment (joakim-hove) [inline devel/libenkf/include/ert/enkf/hook_workflow.h]: This enum defintion can (hopefully?) go to the manager implementation.

File: devel/libenkf/src/hook_manager.c [hunk 6]
```diff
@@ -36,10 +36,8 @@
 
 
 struct hook_manager_struct {
-  char                   * path;
-  workflow_type          * workflow;
-  ert_workflow_list_type * workflow_list;
```
Comment (joakim-hove) [inline devel/libenkf/src/hook_manager.c]: The workflow_list is a shared resource and should be here?

File: devel/libenkf/src/hook_workflow.c (hunk 4 above)
Comment (arielalmendral) [inline devel/libenkf/src/hook_workflow.c]: Yes, the path is to support the old QC_PATH. I was under the impression that we wanted to maintain the old QC functionality as it was.

File: devel/libenkf/src/hook_workflow.c (hunk 3 above)
Comment (arielalmendral) [inline devel/libenkf/src/hook_workflow.c]: The idea here is that a manager is a collection of hooks and that each hook should know when it should be run.

File: devel/libenkf/include/ert/enkf/hook_workflow.h (hunk 5 above)
Comment (arielalmendral) [inline devel/libenkf/include/ert/enkf/hook_workflow.h]: Yes, probably.

File: devel/libenkf/src/hook_manager.c [hunk 7]
```diff
@@ -36,10 +36,8 @@
...
-  runpath_list_type      * runpath_list;
+  hook_workflow_type     * hook_workflow;
+  hook_workflow_type     * post_hook_workflow;
```
Comment (joakim-hove) [inline devel/libenkf/src/hook_manager.c]: I think these should be container types - probably `vector_type`. Maybe also a good idea to have a separate container manage the storage of the workflows, then the other lists can have the same reference multiple times.

**Comment by joakim-hove** [comment]
All the internal tests passed!

**Comment by joakim-hove** [comment]
Sorry this has taken so long - merging.

**Comment by arielalmendral** [comment]
Thanks!

**Pull request merged**